### PR TITLE
Instance import, part 1

### LINKED ISSUE #310: Convert USD instances to maya instances on import
**Is your feature request related to a problem? Please describe.**
Occasionally, we want to import a USD scene that uses instancing into maya, and be able to preserve those instances.

**Describe the solution you'd like**
Have a setting to allow maya to automatically convert usd-instances to maya-instances on import.

**Additional context**
So, we've actually already implemented this, and it works and has been used in production, though only infrequently.  Was updating some of our branches recently, and figured I might as well push it public, in case someone else is interested:

https://github.com/LumaPictures/maya-usd/tree/pr/pxr/maya_importing_instances

Mostly, I think it needs tests added before I think it would be ready for a PR; wasn't sure if I should create this as a WIP PR, or an issue...

### LINKED ISSUE #340: [MAYA-105820] Exported Instances are not importing
**Describe the bug**
Exported Instances as USD are not importing at all. Stage works just fine.

**Steps to reproduce**
1. Create instances
2. Export as USD
3. Import USD

**Expected behavior**
I should be able to import Instances as Maya native and be able to edit them.

**Specs (if applicable):**
 - Windows 10
- Maya Beta PR113

### LINKED ISSUE #723: Models with a parent xform flagged as instanceable = true wont import into maya
When importing a usd containing an mesh with an xform as a parent, and that parent has the flag instanceable = true the model wont import.

This is using file->import->usd.

Jerry Gamache has an example fir to illustrate this behavior.

Thanks,
Koen

### PR DESCRIPTION
**Pull request by JGamache-autodesk**
Brings in the original @sirpalee / @pmolodo instancing code from the following [Luma branch](https://github.com/LumaPictures/maya-usd/tree/pr/pxr/maya_importing_instances).

This code adds an Instance Mode option on import to either build instances or flatten the scene.

**Updates:**
- Merged with latest dev
- masters are loaded recursively and on demand, allowing import of instances inside other instances.
- added unit tests for instance imports based on USD instancing scenarios by @pmolodo 

**Bug fixes:**
- Saves the Maya shape name instead of using the generic name "Shape"
- XformOp errors on import could result in geometries with (0, 0, 0) scaling. Will now default to (1, 1, 1).

Fixes #340, fixes #723, fixes #310

**Still todo:**
The following issues have been found and will likely be fixed in a follow-up update:
- Extra transforms get inserted between the instantiation transform and the instanced shape (also happens for regular non-instanced hierarchies where shapes in the middle of the hierarchy also gain a transform).
- Instanced hierarchies will export as piecewise instances. A transform where all children are instances should generate a single USD master.

### CODE REVIEW

File: test/lib/usd/translators/testUsdImportInstances.py
Comment (mattyjams) [inline test/lib/usd/translators/testUsdImportInstances.py:68]: Ahh... so maybe this explains the missing type names in the USD above. Prims with the `Scope` type name are different from the typeless prims we have at the moment though. Do we really mean `Scope` here and we should add that type name above, or should this read typeless here instead?

File: test/lib/usd/translators/testUsdImportInstances.py
Comment (mattyjams) [inline test/lib/usd/translators/testUsdImportInstances.py:86]: I think this setting of `primPath` can be removed, since that should be the default? Similarly, `instanceMode=buildInstances` looks like it was made the default as well, but I'm ok with that staying since I think it makes the intent of the test clear.

File: test/lib/usd/translators/testUsdImportInstances.py
Comment (mattyjams) [inline test/lib/usd/translators/testUsdImportInstances.py:83]: All of these tests do a File->New. Maybe move this to `setUp()`?

**Review by mattyjams — COMMENTED** [review]
I think I'll have to defer to @pmolodo for the final approval since he's much more knowledgable on this subject than I am, but it generally looks good to me. I had just a few mostly minor notes.

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (pmolodo) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:267]: Agreed, I think we should eliminate the stage args.  To be fair - they were initially introduced in Pal's original commit.  But yeah, I think their introduction is a step backward, so we should eliminate them.
@JGamache-autodesk, if you're ok with it, I can push a cleaned up rebase that doesn't introduce them...

**Review by pmolodo — COMMENTED** [review]
So, I don't have too much additional to add to what @mattyjams said - mostly looks good.

My one concern is that, if we do release as is, and later want to support "cleaning up" the "extra" Scope args, it will obviously change the resulting maya scene.  If we release this as is, people may become dependent on the behavior implemented in this PR... and I'd hate to lock us into having the unnecessary xforms.  Are we ok with making a potentially breaking change to eliminate them in the future?

File: test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda
Comment (mattyjams) [reply test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda:69]: Cool, thanks @pmolodo. That all definitely makes sense to me, and I'm fine with leaving it as `InstanceSources`. I think I had that gut reaction just seeing the USD before I looked at the Python test script, and the intention is made a bit clearer in the docstrings there.

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (pmolodo) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:267]: I pushed a fixed-up version here:

https://github.com/LumaPictures/maya-usd/tree/pr/pxr/maya_importing_instances

File: test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda
Comment (pmolodo) [reply test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda:69]: I guess the other thing that isn't clear from this commit is that the name `InstanceSources` isn't an arbitrarily chosen name for the scope in these sample files, but the "standard" name that Maya creates when *exporting* instances:

https://github.com/Autodesk/maya-usd/blob/fab6650e8149f09fec1df28c6de0a5ae10c04d88/lib/mayaUsd/fileio/writeJobContext.cpp#L83

The more I think about it though - the more I think we should change this name to something more maya specific.  That way we can feel safer about treating it specially later... and if we are going to change it, better sooner rather than later.  (And it should be a very easy change to make...)

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (JGamache-autodesk) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:267]: Got it. Thanks!

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (JGamache-autodesk) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:443]: Not sure about that one. With the current code, using `instanceMode=anything_but_flatten_or_build` would end up not loading any instance, which was the old behavior before this patch.

Is that a valid mode that should be added to the Import options? I would call it "ignore"...

File: test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda
Comment (JGamache-autodesk) [reply test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda:69]: Would that also affect the hardcoded "InstanceSources" in the AnimalLogic plugin?
plugin\al\lib\AL_USDMaya\AL\usdmaya\fileio\Export.cpp:296

File: test/lib/usd/translators/testUsdImportInstances.py
Comment (JGamache-autodesk) [reply test/lib/usd/translators/testUsdImportInstances.py:86]: Isn't primPath the `defaultPrim`, which would be pCube1 in the test scenes?

**Review by pmolodo — APPROVED** [review]
Looks good to me.  I'd still like to change the name of "InstanceSources" to be more maya-specific, but I suppose we can make that a separate PR if you'd prefer.

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (JGamache-autodesk) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:463]: With the special builds we use, we get asserts about incorrect children counts on file->new because the instance count and children count differ. It is not something you will see with a shipped Maya cut.

File: lib/mayaUsd/fileio/jobs/readJob.cpp
Comment (mattyjams) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:435]: No, I don't think it makes sense to preserve the behavior where instances in USD are ignored. If I'm reading the changes in jobArgs.cpp correctly, I think we'll fallback to building instances if instanceMode was given something bogus?

My main thought here was just whether we could re-use the `buildInstances` boolean from above. No big deal though.

File: test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda
Comment (mattyjams) [inline test/lib/usd/translators/UsdImportInstancesTest/ComplexScenario_optimized.usda:8]: One small nit-pick with the type names here, but it looks like we ended up with type names everywhere. I would think the majority of the time when you have a prim referencing other prims, you likely want a typeless def for the prim doing the referencing. I *think* in most cases below, the type names between the referencer and the referenced agree, so I don't think it's a practical issue here. Typeless defs let the type of the referenced prim shine through to the referencer whatever that type may be, so you don't have to match the types yourself. Specifying a different type on the referencer will actually change the type of the composed prim, which is pretty rarely what you intend.

**Review by mattyjams — APPROVED** [review]
Looks good to me!

**Pull request merged**
